### PR TITLE
Create connectBackboneToReact and BackboneProvider

### LINKED ISSUE #2: RFC: Improved connectBackboneToReact and BackboneProvider
##  How do I RFC?

* Read the [Problem](#1-problem) and [Goals](#2-goals), then the proposed solution. 
* If you like the [Proposed Solution](#3-proposed-solution), give a :+1: or an LGTM.
* If you dislike the proposed solution, check out the [alternate solutions](#4-alternate-solutions).
    * If you like one of them, propose one of them and explain why.
    * If you sort of like one of them, propose a change to it.
    * If you dislike all of them, propose a new alternate solution.
* If you'd rather this be something we don't regulate, voice that.

Statuses for RFCs are:
* Discussion
* Adopted
* Not Regulated

## Contents

* [1. Problem](#1-problem)
* [2. Goals](#2-goals)
* [3. Proposed Solution](#3-proposed-solution)
* [4. Alternate Solutions](#4-alternate-solutions)

## 1. Problem

In order to use React Components within Backbone we need to accomplish two things:

1. Be able to render a React Component within Backbone. (This is addressed in a PR).

  After the PR is merged we will be able to render a React Component within a Marionette View like this:

  ```javascript
  class Greeting extends Component {
    render() {
      return (
        <div>
          Hello!
        </div>
      );
    }
  }

  const MyView = mixInto(Marionette.ItemView)(
      viewWithReact,
  ).extend({
      template: () => `
        <div>
            <header>
                <h1>Welcome!</h1>
            </header>

            <div class="react"></div>
        </div>
      `,

      initialize: function(options) {
      
        this.registerReactComponent(
          <Greeting />,
          '.react'
        );
      }
  });

  (new MyView()).render();
  // Rendered!
  ```

2. Be able to share data stored within Backbone Models and Collections between React Components and Backbone Views.

  Consider this case:

  ```javascript
  class Greeting extends Component {
    render() {
      return (
        <div>
          Hello {this.props.name}!
        </div>
      );
    }
  }

  const MyView = mixInto(Marionette.ItemView)(
      viewWithReact,
  ).extend({
      template: () => `
        <div>
            <header>
                <h1>Welcome!</h1>
            </header>

            <div class="react"></div>
        </div>
      `,

      initialize: function(options) {
        const model = new Backbone.Model({
          name: 'Harry',
        });

        // We could do this:

        this.registerReactComponent(
          <Greeting name={model.get('name')} />,
          '.react'
        );

        // However if we then do:

        model.set('name', 'The Loud One');

        // Then the React Component is out of sync with the data.

        // This can be fixed via:

        model.on('change:name', () => {
          this.registerReactComponent(
            <Greeting name={model.get('name')} />,
            '.react'
          );

          this.render();
        });

        // But that quickly gets cumbersome.

        // Alternatively we could fix it like:
        this.registerReactComponent(
          <Greeting model={model} />,
          '.react'
        );

        // And within the Greeting component set up our event handlers directly,
        // but again that gets cumbersome to repeat that boilerplate every time.

        // Additionally a React Component should not have a direct dependence on a Model.
        // We should aim to decouple.
      }
  });

  (new MyView()).render();
  ```

This RFC addresses the second issue.

## 2. Goals

1. Retain one true source of truth. That being, the data stored within the Models and Collections is what all UI should use to render.
2. Create an abstraction between Backbone Models/Collections and React Components. This is to prevent a React Component from being directly dependent on a Model/Collection.
3. Keep the UI in sync with the data in the Models/Collections. Such that whenever the Model/Collection changes the React Component re-renders.
4. Create an API that is as frictionless as possible. This is to encourage usage.

## 3. Proposed Solution

An initial solution has been created via [connect-backbone-to-react](https://github.com/mongodb-js/connect-backbone-to-react) package, however the API has room for improvement before we mark it as `1.0.0`.

This solution uses the [Higher-Order Components](https://facebook.github.io/react/docs/higher-order-components.html) (HoCs) pattern. This is the solution the React community has moved to after finding mixins hard to maintain. The end goal is the same as mixins, sharing functionality, however HoCs are easier to reason about. Refer to that document for further explanation.

`connectBackboneToReact` creates a HoC. The HoC returned:

* Manages creating Model/Collection eventListeners so that when the model changes, the enclosed component is re-rendered.
* Handles cleaning up eventListeners when the component is unmounted.
* Maps Models/Collections to `props` that are given to your component.
* Passes any additional props given to the HoC to the WrappedComponent. This allows you to still pass `props` that change non-data related behaviors, such as `showPlaceholder={false}`.

Ultimately we have four inputs required to correctly create our HoC:

- `modelsMap` - this defines the map of Models/Collections to local names.
  - Example: `{ user: userModel, settings: settingsCollection }`

- `mapModelsToProps` - this is function that maps the values from `modelsMap` to `props` that will be given to our WrappedComponent.
  - Example:
    ```javascript
    function({ user, settings }) {
      return {
        name: user.get('name'),
        changeName: (newName) => user.set('name', newName)
      };
    };
    ```

- `options` - general options object to manipulate how the wrapper behaves.
  - `options.debounce`: Should our event handler function be wrapped in a debounce function. By default it's _false_.
  - `options.events`: Custom map of events our HoC should subscribe to. By default the _all_ event is used for each Model/Collection.
  - `options.modelTypes`: Define what Model/Collection types the HoC expects to be given in the `modelsMap`. This is patterned off of [React's PropTypes](https://facebook.github.io/react/docs/typechecking-with-proptypes.html).

-  `WrappedComponent` - our own React Component that is going to be wrapped.

The full usage would look like:

```javascript
// ComponentA.js
const { connectBackboneToReact } = require('connect-backbone-to-react');

class ComponentA extends Component {
  render() {
    return (
      <div>
        Hello {this.props.name}!
        <div
          onClick={() => {
            this.props.setName(Math.random());
          }}
      </div>
    );
  }
}

const mapModelsToProps = ({ user }) => {
  return {
    name: user.get('name'),
    setName: (newName) => user.set('name', newName)
  };
},
const options = {
  debounce: true,
  events: {
    user: ['change', 'reset'],
  },
  modelTypes: {
    user: Backbone.Model,
  },
};

module.exports = connectBackboneToReact(
  mapModelsToProps,  
  options
)(ComponentA);

// MyView.js

// ComponentA is the connected HoC.
const ComponentA = require('./ComponentA.js');

const MyView = mixInto(Marionette.ItemView)(
  viewWithReact,
).extend({
  template,

  initialize: function(options) {
    const model = new Backbone.Model({
      name: 'Harry',
    });

    const modelsMap = {
      user: model,
    };

    this.registerReactComponent(
      <ComponentA models={modelsMap} />,
      '.react'
    );
  }
});
```

---

That covers the basic use case of rendering one connected React Component. However as we incrementally migrate more Backbone Views to React we'll end up with a tree of React Components. So what happens if we have a tree of connected React Components?

```javascript
class ComponentA extends Component {
  render() {
    return (
      <div>
        <ConnectedComponentB>
          <ConnectedComponentC />
        </ConnectedComponentB>
      </div>
    );
  }
}
```

We shouldn't pass that `modelsMap` object from one component to another. Instead we can take inspiration from react-redux's [Provider component](https://github.com/reactjs/react-redux/blob/master/docs/api.md#provider-store).

This allows us to share the `modelsMap` via [React's Context feature](https://facebook.github.io/react/docs/context.html).

Usage would look like:

```javascript
// ComponentA is the connected HoC.
const ComponentA = require('./ComponentA.js');
const { BackboneProvider } = require('connect-backbone-to-react');

const MyView = mixInto(Marionette.ItemView)(
  viewWithReact,
).extend({
  template,

  initialize: function(options) {
    const model = new Backbone.Model({
      name: 'Harry',
    });

    const modelsMap = {
      user: model,
    };

    this.registerReactComponent(
      <BackboneProvider models={modelsMap}>
        <ComponentA />
      </BackboneProvider>,
      '.react'
    );
  }
});
```

`BackboneProvider` shares the `modelsMap` via `context`. Every HoC connected to Backbone would then retrieve the `modelsMap` via `context`. 

### What should the `props` name be that the `modelsMap` is given to?

Right now you pass in `modelsMap` through the `models` prop:

```javascript
<ComponentA models={modelsMap} />
```

Alternatively: 

```javascript
<ComponentA data={modelsMap} />
```

### Where should we put `modelTypes`?

We have the proposed solution, here's an alternative which mirrors the style that `PropTypes` are added to React components.

```javascript
const GreetingConnected = connectBackboneToReact(
  mapModelsToProps,
  options,
)(Greeting);

GreetingConnected.modelTypes = {
  user: instanceOf(UserModel),
  allUsers: instanceOf(UserCollection),
};
```

### PR DESCRIPTION
**Pull request by hswolff**
This resolves #2.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -13,6 +13,7 @@
     }
   },
   "rules": {
+    "eqeqeq": ["error", "smart"],
```
Comment (jrbalsano) [inline .eslintrc]: Sorry! I have to speak up now that I don't like this rule :P I think it's simpler when we have to spell it out for null and undefined. That would be consistent with our other repos too.

File: README.md [hunk 2]
```diff
@@ -25,46 +27,77 @@ class MyComponent extends React.Component {
...
-        user.set('laughs', newVal);
-      },
-    };
-  }
+  // Define what modelTypes you expect to be contained on your `modelsMap` object.
+  // Useful for validating that you'll be given what model type you expect.
+  // Uses instanceof, and throws an error if instanceof returns false.
+  // By default no modelTypes are defined.
+  modelTypes: {
+    user: Backbone.Model,
```
Comment (jrbalsano) [inline README.md]: it might be good to demonstrate putting something other than `Backbone.Model` here. Maybe changing out `userModel` from using `Backbone.Model` to using something like `const User = Backbone.Model.extend({ ... });`

File: README.md [hunk 3]
```diff
@@ -25,46 +27,77 @@ class MyComponent extends React.Component {
...
-    // that will cause your React Component to re-render.
-    // By default it's ['all'] for every Model and Collection given.
-    events: {
-      user: ['change:name', 'change:laughs'], // You can disable listening to events by passing in `false`.
-    },
-  }
-)(MyComponent);
+const { connectBackboneToReact } = require('connect-backbone-to-react');
+
+// Create our Connected HoC.
```
Comment (jrbalsano) [inline README.md]: I think this is the first use of HoC here, so it might be good to say "Higher order Component (HOC)" (As a side note it looks like the React documentation uses HOC all caps)

File: README.md [hunk 4]
```diff
@@ -25,46 +27,77 @@ class MyComponent extends React.Component {
...
+### BackboneProvider
+
+Alternatively you might have a tree of connected Components. We shouldn't pass that `modelsMap` object from one component to another. Instead we can take inspiration from [react-redux's Provider component](https://github.com/reactjs/react-redux/blob/master/docs/api.md#provider-store).
+
+```javascript
+const { BackboneProvider } = require('connect-backbone-to-react');
 
 ReactDOM.render(
-  <WrappedComponent />,
+  // Pass the modelsMap to the BackboneProvider via the models prop.
```
Comment (jrbalsano) [inline README.md]: It might be good to demonstrate that `modelsMap` here is the same `modelsMap` as the previous section by repeating it. I think otherwise it might be unclear how to use the `BackboneProvider`

File: lib/backbone-provider.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+const { Component, PropTypes, Children } = require('react');
+
+class BackboneProvider extends Component {
```
Comment (jjgonecrypto) [inline lib/backbone-provider.js:3]: boo to 2 spaces 👻

Comment (jrbalsano) [reply lib/backbone-provider.js:3]: Silly thing but I would actually also appreciate if we went with 4 spaces again because it would be consistent with our other repos. If we are going to depart from that I think maybe it should be brought up with the chapter?

File: lib/connect-backbone-to-react.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+      }
+    });
+  }
+
+  return function createWrapper(WrappedComponent) {
+    const wrappedComponentName = WrappedComponent.displayName
+      || WrappedComponent.name
+      || 'Component';
+
+    const displayName = getDisplayName(wrappedComponentName);
```
Comment (jrbalsano) [inline lib/connect-backbone-to-react.js:63]: Is this used by the debugging tools?

**Review by jrbalsano — APPROVED** [review]
Mostly nits and questions. Looks really great! The one thing that I have the most trouble with (expected) is the connectBackboneToReact file though.It may just be because that kind of code is unfamiliar to me, but I find it just a little hard to follow. There's a very large use of closures combined with an actual class definition with it's own state that I feel combine to make for a very confusing file. I wonder if we could reduce some of the closure usage in there by currying or something and then extract the functionality into smaller files?

File: lib/backbone-provider.js (hunk 5 above)
Comment (hswolff) [reply lib/backbone-provider.js:3]: To preface, my editor defaults to 2 spaces on new projects, so this wasn't done intentionally. Although I do prefer 2 spaces.

Question though, the rest of the mongodb-js repos use 2 spaces, internally we use 4 spaces.

Happy to do whatever but just wanted to raise that.

File: .eslintrc (hunk 1 above)
Comment (hswolff) [reply .eslintrc]: Fair enough.

File: README.md (hunk 3 above)
Comment (hswolff) [reply README.md]: Nice catch!

File: lib/connect-backbone-to-react.js (hunk 6 above)
Comment (hswolff) [reply lib/connect-backbone-to-react.js:63]: Yes sir!

**Comment by hswolff** [comment]
> There's a very large use of closures combined with an actual class definition with it's own state that I feel combine to make for a very confusing file. I wonder if we could reduce some of the closure usage in there by currying or something and then extract the functionality into smaller files?
Show outdated

Currying would just move the closure usage elsewhere.

I definitely agree that this file could grow, and if it does I think looking into breaking up functionality into smaller functions could be looked into.

The file is still small however, so I think breaking it up right now would add complexity rather than reduce it.

File: lib/backbone-provider.js (hunk 5 above)
Comment (jjgonecrypto) [reply lib/backbone-provider.js:3]: I think we should align with `mongodb-js` in this instance...

File: lib/backbone-provider.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  constructor(props, context) {
+    super(props, context);
+
+    this.modelsMap = props.models;
```
Comment (jjgonecrypto) [inline lib/backbone-provider.js]: [np] why attach this to the instance when it's already in `props`?

File: lib/connect-backbone-to-react.js [hunk 8]
```diff
@@ -0,0 +1,147 @@
+const hoistStatics = require('hoist-non-react-statics');
+const { Component, PropTypes, createElement } = require('react');
+const debounceFn = require('lodash.debounce');
+const isArray = require('lodash.isarray');
+
+function getDisplayName(name) {
```
Comment (jjgonecrypto) [inline lib/connect-backbone-to-react.js]: [np] any reason to not use arrow syntax?

File: lib/connect-backbone-to-react.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+function defaultMapModelsToProps(models) {
+  return Object.keys(models).reduce((acc, modelKey) => {
+    const model = models[modelKey];
+    acc[modelKey] = model.toJSON();
+    return acc;
+  }, {});
+}
+
+module.exports = function connectBackboneToReact(
```
Comment (jjgonecrypto) [inline lib/connect-backbone-to-react.js:17]: ibid

File: lib/connect-backbone-to-react.js [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+  function getEventNames(modelName) {
+    let eventNames = events[modelName];
+
+    // Allow turning off event handlers by setting events to false.
+    if (eventNames === false) {
+      return [];
+    }
+
+    if (!isArray(eventNames)) {
```
Comment (jjgonecrypto) [inline lib/connect-backbone-to-react.js]: You can use `Array.isArray` - IE9 and above

File: lib/connect-backbone-to-react.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+      eventNames = [];
+    }
+
+    if (eventNames.length === 0) {
```
Comment (jjgonecrypto) [inline lib/connect-backbone-to-react.js]: is it worth checking this case? instead of just initializing it to `['all']` if it's not an Array above? Would there be a case where the user would set the `events[modelName]` property to an empty array?

File: lib/connect-backbone-to-react.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+  }
+
+  function validateModelTypes(modelsMap) {
+    return Object.keys(modelTypes).forEach(modelKey => {
+      const ModelConstructor = modelTypes[modelKey];
+      const modelInstance = modelsMap[modelKey];
+
+      const isInstanceOfModel = modelInstance instanceof ModelConstructor;
+      if (!isInstanceOfModel) {
+        throw new Error(`"${modelKey}" model found on modelsMap does not match type required.`);
```
Comment (jjgonecrypto) [inline lib/connect-backbone-to-react.js:53]: [np] Maybe add actual & expected in `Error` so the dev can easily see which is which.

File: lib/connect-backbone-to-react.js [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+      || WrappedComponent.name
+      || 'Component';
+
+    const displayName = getDisplayName(wrappedComponentName);
+
+    class Connect extends Component {
+      constructor(props, context) {
+        super(props, context);
+
+        this.modelsMap = props.models || context.models;
```
Comment (jjgonecrypto) [inline lib/connect-backbone-to-react.js]: it's a bit confusing having `modelsMap` and `mapModelsToProps`. Could you just have `this.models` instead?

File: lib/connect-backbone-to-react.js [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        });
+      }
+
+      createNewProps() {
+        this.setState(this.mapModelsToProps());
+      }
+
+      componentWillUnmount() {
+        if (debounce) {
+          this.createNewProps.cancel();
```
Comment (jjgonecrypto) [inline lib/connect-backbone-to-react.js:101]: 👍

File: lib/connect-backbone-to-react.js [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+        Object.keys(this.modelsMap).forEach(mapKey => {
+          const model = this.modelsMap[mapKey];
+
+          getEventNames(mapKey).forEach(name => {
+            model.off(name, this.createNewProps, this);
+          });
+        });
+      }
+
+      mapModelsToProps() {
```
Comment (jjgonecrypto) [inline lib/connect-backbone-to-react.js]: is this helper function really necessary?

File: test/backbone-provider.test.js [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    });
+
+    it('passes mapped models and collections as properties to wrapped component', function() {
+      assert.deepEqual(componentStub.props().user, userModel.toJSON());
+      assert.deepEqual(componentStub.props().coll, userCollection.toJSON());
+    });
+
+    it('every connected Component is given the same "modelsMap"', function() {
+      assert(
+        wrapper.find('.name').everyWhere(n => n.text() === 'Harry')
```
Comment (jjgonecrypto) [inline test/backbone-provider.test.js:98]: perhaps also assert that there's two instances of Harry found? Ensure the child has correctly gotten its data? (Though I'm guessing `this.props.user.name` would throw in the child render if the childContext wasn't correctly set?)

File: test/connect-backbone-to-react.test.js [hunk 17]
```diff
@@ -5,16 +5,27 @@ const _ = require('lodash');
...
 describe('connectBackboneToReact', function() {
   let sandbox;
   let wrapper;
   let stub;
+  let modelsMap;
+  let mapModelsToProps;
+
   let userModel;
+  const UserModel = Model.extend({
+  });
```
Comment (jjgonecrypto) [inline test/connect-backbone-to-react.test.js]: [np] is the newline really necessary?

File: test/connect-backbone-to-react.test.js [hunk 18]
```diff
@@ -151,19 +162,22 @@ describe('connectBackboneToReact', function() {
         assert.equal(getEventHandlersSetCount(model, wrapperInstance), 0);
       });
     });
+
```
Comment (jjgonecrypto) [inline test/connect-backbone-to-react.test.js]: above on the test `updates properties when model and collections change`, can you add an assertion that ensures `'Banana'` is rendered?

File: test/connect-backbone-to-react.test.js [hunk 19]
```diff
@@ -53,20 +64,20 @@ describe('connectBackboneToReact', function() {
   beforeEach(function() {
     sandbox = sinon.sandbox.create();
 
-    userModel = new Model({
+    userModel = new UserModel({
```
Comment (jjgonecrypto) [inline test/connect-backbone-to-react.test.js:53]: Above on L49 you are using a backbone internal property `_events`. An alternative is to stub out `on` and `off` and manage it that way.

**Review by jjgonecrypto — COMMENTED** [review]
Mostly nps, though a few comments on the tests (some older stuff I'd missed)

File: lib/connect-backbone-to-react.js (hunk 8 above)
Comment (hswolff) [reply lib/connect-backbone-to-react.js]: Original goal of this was to have it be as vanilla JS as possible, with the hopes of not needing any transpiling. However as I coded more that goal was removed, and this is a relic of that original desire.

I don't really care either way. If you'd like me to change let me know!

File: lib/connect-backbone-to-react.js (hunk 11 above)
Comment (hswolff) [reply lib/connect-backbone-to-react.js]: Very weird...I tried to remember my reasoning for this and I can't. Weird mind glitch. Nice catch.

Comment (hswolff) [reply lib/connect-backbone-to-react.js]: Oh I remember! If a user sets the value of the event to an empty array this guarantees that the 'all' event will be added. Which in retrospect is really bad behavior. 

i.e.

```javascript
const options = {
  events: {
    user: []
  },
};
```

That would be very unexpected behavior. Going to fix that.

File: lib/connect-backbone-to-react.js (hunk 12 above)
Comment (hswolff) [reply lib/connect-backbone-to-react.js:53]: Ideally I could show the constructor name but I tried to figure out a way to sniff for that value and couldn't figure it out.

image

File: test/connect-backbone-to-react.test.js (hunk 19 above)
Comment (hswolff) [reply test/connect-backbone-to-react.test.js:53]: Oh damn, that's much better. Much more work involved but definitely more robust.

File: test/backbone-provider.test.js (hunk 16 above)
Comment (hswolff) [reply test/backbone-provider.test.js:98]: This was done but in the line above.

**Comment by hswolff** [comment]
Ready for round 2 of review!

Also in the latest push I added `standard-version` that automates creating `CHANGELOG.md` files (so long as you follow their conventions).

When you run `npm run release` it'll automatically increment the package version according to semver and update the `CHANGELOG.md` file.

When I ran it the first time it created the following:

---

# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.6.0"></a>
# 0.6.0 (2017-02-24)


### Bug Fixes

* Actually add packages we depend on to our dependencies list. ([3783799](https://github.com/mongodb-js/connect-backbone-to-react/commit/3783799))
* PR feedback from Justin ([9f3b24c](https://github.com/mongodb-js/connect-backbone-to-react/commit/9f3b24c))


### Features

* Add modelTypes as an option you can set. ([1eac09f](https://github.com/mongodb-js/connect-backbone-to-react/commit/1eac09f))
* Create BackboneProvider ([e1c3703](https://github.com/mongodb-js/connect-backbone-to-react/commit/e1c3703))
* Initial import of all functionality! ([0b16681](https://github.com/mongodb-js/connect-backbone-to-react/commit/0b16681))
* Move react to peerDep ([11c23ce](https://github.com/mongodb-js/connect-backbone-to-react/commit/11c23ce))
* Update API for connectBackboneToReact to require passing modelsMap through as a prop. ([434e805](https://github.com/mongodb-js/connect-backbone-to-react/commit/434e805))

File: lib/connect-backbone-to-react.js (hunk 12 above)
Comment (jjgonecrypto) [reply lib/connect-backbone-to-react.js:53]: Try `UserModel.prototype` ;)

File: test/backbone-provider.test.js (hunk 16 above)
Comment (jjgonecrypto) [reply test/backbone-provider.test.js:98]: you mean with the props? I meant within the DOM

File: test/connect-backbone-to-react.test.js (hunk 19 above)
Comment (jjgonecrypto) [reply test/connect-backbone-to-react.test.js:53]: are you going to do this?

File: test/connect-backbone-to-react.test.js (hunk 18 above)
Comment (jjgonecrypto) [reply test/connect-backbone-to-react.test.js]: ping

File: lib/connect-backbone-to-react.js (hunk 12 above)
Comment (hswolff) [reply lib/connect-backbone-to-react.js:53]: image

File: test/backbone-provider.test.js (hunk 16 above)
Comment (hswolff) [reply test/backbone-provider.test.js:98]: Sorry no, I meant I added a new test assertion above which is why this comment isn't showing as outdated. 

```
assert.equal(wrapper.find('.name').length, 2);
```

File: test/connect-backbone-to-react.test.js (hunk 19 above)
Comment (hswolff) [reply test/connect-backbone-to-react.test.js:53]: Yeah it's done, just not here so doesn't show as outdated.

File: test/connect-backbone-to-react.test.js (hunk 18 above)
Comment (hswolff) [reply test/connect-backbone-to-react.test.js]: Whoops, thanks for the ping!

File: README.md [hunk 20]
```diff
@@ -4,9 +4,12 @@
...
 ## Example
 
+### connectBackboneToReact
+
 ```javascript
-const userModel = new Backbone.Model({ name: 'Harry', laughs: true });
-const userCollection = new Backbone.Collection([userModel]);
+const UserModel = Backbone.Model.extend();
+const userInstance = new UserModel({ name: 'Harry', laughs: true });
+const userCollection = new Backbone.Collection([userInstance]);
```
Comment (jrbalsano) [inline README.md]: What do you think about create a UserCollection with `{ model: UserModel }` and userCollectionInstance? That would let you demonstrate multiple modelTypes (which I usually find that helpful for demonstrating/testing hypotheses/patterns when reading documentation)

File: README.md [hunk 21]
```diff
@@ -25,46 +28,83 @@ class MyComponent extends React.Component {
...
-  }
-)(MyComponent);
+const { connectBackboneToReact } = require('connect-backbone-to-react');
+
+// Create our Connected Higher order Component (HOC).
+const MyComponentConnected = connectBackboneToReact(mapModelsToProps, options)(MyComponent);
+
+// Map your Backbone Model and Collections to names that will be provided to
+// your mapModelsToProps function.
+const modelsMap = {
```
Comment (jrbalsano) [inline README.md]: I think I'd break the codeblocks here to be able to say something like "Now that you've defined you're connected HOC you can use it!" A visual clue demonstrating the difference between what is meant to be setup and what is meant to be use would be helpful for me.

File: lib/backbone-provider.js [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+  }
+
+  render() {
+    return Children.only(this.props.children);
+  }
+}
+
+BackboneProvider.propTypes = {
+  models: PropTypes.object,
+  children: PropTypes.element.isRequired,
```
Comment (jrbalsano) [inline lib/backbone-provider.js:17]: Does this mean a Provider can only have one direct child element?

**Review by jrbalsano — APPROVED** [review]
Just in case you thought you were waiting for me on this I just have a bunch of documentation critiques :P This LGTM! I wanna use it already!

File: README.md (hunk 20 above)
Comment (hswolff) [reply README.md]: 👍

File: lib/backbone-provider.js (hunk 22 above)
Comment (hswolff) [reply lib/backbone-provider.js:17]: Correct. This is mimicked from react-redux, no other reason beside that.

File: README.md (hunk 21 above)
Comment (hswolff) [reply README.md]: 👍

File: lib/connect-backbone-to-react.js (hunk 12 above)
Comment (jjgonecrypto) [reply lib/connect-backbone-to-react.js:53]: Ugh - I'm curious what magic the Chrome console is pulling to log the type there.

**Review by jjgonecrypto — APPROVED** [review]
